tsfc/driver.py

```diff
@@ -267,21 +269,104 @@ def name_multiindex(multiindex, name):
+
+    'contravariant piola' mapping for g:
+
+      G(X) = det(J) K g(x)   i.e  G_i(X) = det(J) K_ij g_j(x)
+
+    'covariant piola' mapping for g:
+
+      G(X) = J^T g(x)          i.e  G_i(X) = J^T_ij g(x) = J_ji g_j(x)
+
+    'double covariant piola' mapping for g:
+
+      G(X) = J^T g(x) J     i.e. G_il(X) = J_ji g_jk(x) J_kl
+
+    'double contravariant piola' mapping for g:
+
+      G(X) = det(J)^2 K g(x) K^T  i.e. G_il(X)=(detJ)^2 K_ij g_jk K_lk
+
+    If 'contravariant piola' or 'covariant piola' are applied to a
+    matrix-valued function, the appropriate mappings are applied row-by-row.
+    """
+
+    # Find out which mapping to apply
+    try:
+        mapping, = set(to_element.mapping())
+    except ValueError:
+        raise NotImplementedError("Don't know how to interpolate onto zany spaces, sorry")
+
+    mesh = expression.ufl_domain()
+    rank = len(expression.ufl_shape)
+    if mapping == "affine":
+        return expression
+    elif mapping == "covariant piola":
+        J = Jacobian(mesh)
+        *i, j, k = ufl.indices(len(expression.ufl_shape) + 1)
+        expression = ufl.classes.Indexed(expression, ufl.classes.MultiIndex((*i, k)))
+        return as_tensor(J.T[j, k] * expression, (*i, j))
+    elif mapping == "contravariant piola":
+        mesh = expression.ufl_domain()
+        K = JacobianInverse(mesh)
+        detJ = JacobianDeterminant(mesh)
+        *i, j, k = ufl.indices(len(expression.ufl_shape) + 1)
+        expression = ufl.classes.Indexed(expression, ufl.classes.MultiIndex((*i, k)))
+        return as_tensor(detJ * K[j, k] * expression, (*i, j))
+    elif mapping == "double covariant piola" and rank == 2:
+        mesh = expression.ufl_domain()
+        J = Jacobian(mesh)
+        return J.T * expression * J
+    elif mapping == "double contravariant piola" and rank == 2:
+        mesh = expression.ufl_domain()
+        K = JacobianInverse(mesh)
+        detJ = JacobianDeterminant(mesh)
+        return (detJ)**2 * K * expression * K.T
+    else:
+        raise NotImplementedError("Don't know how to handle mapping type %s for expression of rank %d" % (mapping, rank))
+
+
+def compile_expression_dual_evaluation(expression, to_element, coordinates, interface=None,
+                                       parameters=None, coffee=False):
+    """Compile a UFL expression to be evaluated against a compile-time known reference element's dual basis.
+
+    Useful for interpolating UFL expressions into e.g. N1curl spaces.
 
     :arg expression: UFL expression
-    :arg points: reference coordinates of the evaluation points
+    :arg to_element: A FIAT FiniteElement for the target space
     :arg coordinates: the coordinate function
     :arg interface: backend module for the kernel interface
     :arg parameters: parameters object
     :arg coffee: compile coffee kernel instead of loopy kernel
     """
     import coffee.base as ast
     import loopy as lp
+    from FIAT.functional import PointEvaluation
```

Comment:
Please import at the top of the module.

I think the two imports above are here, because we import neither coffee nor loopy at module load time, so if someone only uses one of those two output formats, they don't need the other one installed as a dependency.